odoo/orm/models.py

```diff
@@ -3416,142 +3412,274 @@ def _check_access(self, operation: str) -> tuple[Self, Callable] | None:
-        for func in self._ondelete_methods:
-            # func._ondelete is True if it should be called during uninstallation
-            if func._ondelete or not self.env.context.get(MODULE_UNINSTALL_FLAG):
-                func(self)
+    @typing.final
+    def _delete_collect_cascade(self) -> list[BaseModel]:
+        cascade_deleted = []
 
-        # TOFIX: this avoids an infinite loop when trying to recompute a
-        # field, which triggers the recomputation of another field using the
-        # same compute function, which then triggers again the computation
-        # of those two fields
-        for field in self._fields.values():
-            self.env.remove_to_compute(field, self)
+        for field in self.env.transaction.registry.many2one_targeting.get(self._name, ()):
+            if field.ondelete != 'cascade':
+                continue
 
-        self.env.flush_all()
+            records_impacted = self._delete_impacted_records(field)
+            if not records_impacted:
+                continue
 
-        cr = self.env.cr
-        Data = self.env['ir.model.data'].sudo().with_context({})
-        Defaults = self.env['ir.default'].sudo()
-        Attachment = self.env['ir.attachment'].sudo()
-        ir_model_data_unlink = Data
-        ir_attachment_unlink = Attachment
+            cascade_deleted.append(records_impacted)
 
-        # mark fields that depend on 'self' to recompute them after 'self' has
-        # been deleted (like updating a sum of lines after deleting one line)
-        with self.env.protecting(self._fields.values(), self):
-            self.modified(self._fields, before=True)
+        return cascade_deleted
 
-        for sub_ids in split_every(cr.IN_MAX, self.ids):
-            records = self.browse(sub_ids)
+    @typing.final
+    def _delete_collect_nullify(self, ignore_records_dict) -> list[tuple[Field, tuple[int]]]:
+        cascade_nulls = []
 
-            cr.execute(SQL(
-                "DELETE FROM %s WHERE id IN %s",
-                SQL.identifier(self._table), sub_ids,
-            ))
+        for field in self.env.transaction.registry.many2one_targeting.get(self._name, ()):
+            if field.ondelete != 'set null':
+                continue
+
+            records_impacted = self._delete_impacted_records(field)
+            if not records_impacted:
+                continue
+
+            model_name = field.model_name
+            if model_name in ignore_records_dict:
+                records_impacted -= ignore_records_dict[model_name]
+                if not records_impacted:
+                    continue
+
+            cascade_nulls.append((field, records_impacted))
+
+        return cascade_nulls
 
+    def _delete_collect_extra(self) -> list[BaseModel]:
+        Data = self.env['ir.model.data'].sudo().with_context({})
+        Defaults = self.env['ir.default'].sudo()
+        Attachment = self.env['ir.attachment'].sudo().with_context(skip_res_field_check=True)
```

Comment:
Nitpick: inline model name later.

```suggestion
        env = self.sudo().env
```